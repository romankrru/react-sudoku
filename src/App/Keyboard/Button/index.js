import React from "react";
import cn from "classnames";

import styles from "./index.module.css";

const Button = (props) => (
	<button
		className={cn(styles.button, {
			[styles.buttonBig]: props.size === "big",
		})}
		onClick={props.onClick}
		disabled={props.disabled}
	>
		{props.children}
	</button>
);

export default Button;
