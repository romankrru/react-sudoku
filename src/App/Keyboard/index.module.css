.keyboard {
    margin: 0 auto;
    width: 600px;
}

.keyboardRow {
    display: flex;
    justify-content: center;
    margin-bottom: 25px;
}