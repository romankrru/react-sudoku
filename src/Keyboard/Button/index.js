import React from "react";
import cn from "classnames";

import styles from "./index.module.css";

const Button = (props) => {
	return (
		<button
			className={cn(styles.button, {
				[styles.buttonBig]: props.size === "big",
			})}
			disabled={props.disabled}
		>
			{props.children}
		</button>
	);
};

export default Button;
