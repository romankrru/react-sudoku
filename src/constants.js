const difficulty = {
	EASY: "EASY",
	MEDIUM: "MEDIUM",
	HARD: "HARD",
};

const modes = {
	NORMAL: "NORMAL",
	CANDIDATE: "CANDIDATE",
};

const EMPTY_CELL = 0;

export { difficulty, modes, EMPTY_CELL };
