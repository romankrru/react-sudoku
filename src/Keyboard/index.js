import React from "react";

import Button from "./Button";
import styles from "./index.module.css";

const Keyboard = (props) => {
	return (
		<div className={styles.keyboard}>
			<div className={styles.keyboardRow}>
				<Button>1</Button>
				<Button disabled>2</Button>
				<Button>3</Button>
				<Button>4</Button>
				<Button>5</Button>
				<Button>6</Button>
				<Button>7</Button>
			</div>
			<div className={styles.keyboardRow}>
				<Button>8</Button>
				<Button>9</Button>
				<Button>âœ•</Button>
				<Button size="big">Normal</Button>
			</div>
		</div>
	);
};

export default Keyboard;
