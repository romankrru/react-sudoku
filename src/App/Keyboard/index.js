import React from "react";
import _ from "lodash";
import T from "prop-types";

import { modes } from "../../constants";
import Button from "./Button";
import styles from "./index.module.css";

const Keyboard = (props) => (
	<div className={styles.keyboard}>
		<div className={styles.keyboardRow}>
			{_.range(1, 8).map((i) => (
				<Button
					key={i}
					candidate={props.mode === modes.CANDIDATE}
					candidateValue={i}
					disabled={props.completedKeys.includes(i)}
					onClick={() => props.onClick(i)}
				>
					{i}
				</Button>
			))}
		</div>
		<div className={styles.keyboardRow}>
			{_.range(8, 10).map((i) => (
				<Button
					key={i}
					candidate={props.mode === modes.CANDIDATE}
					candidateValue={i}
					disabled={props.completedKeys.includes(i)}
					onClick={() => props.onClick(i)}
				>
					{i}
				</Button>
			))}
			<Button onClick={() => props.onClick("Delete")}>âœ•</Button>
			<Button size="big" onClick={() => props.onClick("Mode")}>
				{props.mode === modes.NORMAL ? "Normal" : "Candidate"}
			</Button>
		</div>
	</div>
);

Keyboard.propTypes = {
	completedKeys: T.arrayOf(T.number).isRequired,
	onClick: T.func.isRequired,
	mode: T.string.isRequired,
};

export default Keyboard;
