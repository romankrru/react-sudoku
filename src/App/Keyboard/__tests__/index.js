import React from "react";
import { render, screen } from "@testing-library/react";

import Keyboard from "../index";
import { modes } from "../../../constants";

const renderKeyboard = () => {
	const defaultProps = {
		completedKeys: [],
		onClick: jest.fn,
		mode: modes.NORMAL,
	};

	const view = render(<Keyboard {...defaultProps} />);

	const rerender = (newProps) =>
		view.rerender(<Keyboard {...{ ...defaultProps, ...newProps }} />);

	return { view, rerender };
};

describe("<Keyboard />", () => {
	it("renders 9 digit keys", () => {
		const { view } = renderKeyboard();

		expect(
			[...view.container.querySelectorAll(".button span")].filter((node) =>
				Number.isFinite(Number(node.textContent))
			)
		).toHaveLength(9);

		expect(view.container).toMatchSnapshot();
	});

	it("disables completed keys", () => {
		const { view, rerender } = renderKeyboard();

		expect(
			view.container.querySelectorAll(".button.buttonDisabled")
		).toHaveLength(0);

		rerender({ completedKeys: [1, 2] });

		expect(
			view.container.querySelectorAll(".button.buttonDisabled")
		).toHaveLength(2);

		rerender({ completedKeys: [6, 7, 8, 9] });

		expect(
			view.container.querySelectorAll(".button.buttonDisabled")
		).toHaveLength(4);

		rerender({ completedKeys: [3] });

		expect(
			view.container.querySelectorAll(".button.buttonDisabled")
		).toHaveLength(1);
	});

	it("renders correct mode button", () => {
		const { view, rerender } = renderKeyboard();

		expect(view.getByText("Normal")).toBeInTheDocument();

		rerender({ mode: modes.CANDIDATE });
		expect(view.getByText("Candidate")).toBeInTheDocument();
	});
});
