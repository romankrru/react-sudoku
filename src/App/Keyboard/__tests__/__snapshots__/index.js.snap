// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Keyboard /> renders 9 digit keys 1`] = `
<div>
  <div
    class="keyboard"
  >
    <div
      class="keyboardRow"
    >
      <div
        class="button"
      >
        <span>
          1
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          2
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          3
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          4
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          5
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          6
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          7
        </span>
      </div>
    </div>
    <div
      class="keyboardRow"
    >
      <div
        class="button"
      >
        <span>
          8
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          9
        </span>
      </div>
      <div
        class="button"
      >
        <span>
          âœ•
        </span>
      </div>
      <div
        class="button buttonBig"
      >
        <span>
          Normal
        </span>
      </div>
    </div>
  </div>
</div>
`;
